### PR TITLE
migrates fetch to use sql queries

### PR DESCRIPTION
**Pull request by stephkyou**
This change updates `dolt fetch` to use the appropriate sql engine to generate results.

Related: https://github.com/dolthub/dolt/issues/3922

### CODE REVIEW

File: go/cmd/dolt/dolt.go [hunk 1]
```diff
@@ -453,6 +452,15 @@ func runMain() int {
 		return 1
 	}
 
+	if subcommandName == "fetch" || subcommandName == "pull" || subcommandName == "push" {
```
Comment (macneale4) [inline go/cmd/dolt/dolt.go:455]: move down to line 628.

File: go/cmd/dolt/commands/fetch.go [hunk 2]
```diff
@@ -67,54 +68,69 @@ func (cmd FetchCmd) ArgParser() *argparser.ArgParser {
...
+func constructInterpolatedDoltFetchQuery(apr *argparser.ArgParseResults) (string, error) {
+	var params []interface{}
+
+	var buffer bytes.Buffer
+	var first bool
+	first = true
+
+	buffer.WriteString("call dolt_fetch(")
+
+	writeToBuffer := func(s string) {
```
Comment (macneale4) [inline go/cmd/dolt/commands/fetch.go:113]: I think this function is simple enough that you could just use a strings.join() for this.

File: integration-tests/bats/fetch.bats [hunk 3]
```diff
@@ -0,0 +1,322 @@
+#!/usr/bin/env bats
+load $BATS_TEST_DIRNAME/helper/common.bash
+
+setup() {
+    setup_common
+    TMPDIRS=$(pwd)/tmpdirs
```
Comment (macneale4) [inline integration-tests/bats/fetch.bats:6]: TMPDIRS is very close to the standard TMPDIR. How about FETCHROOT?

File: integration-tests/bats/fetch.bats [hunk 4]
```diff
@@ -0,0 +1,322 @@
...
+    [ "$status" -eq 0 ]
+    [[ "$output" =~ "added table" ]] || false
+
+    run dolt sql -q "show tables as of hashof('origin/main')" -r csv
+    [ "${#lines[@]}" -eq 2 ]
+    [[ "$output" =~ "Table" ]] || false
+    [[ "$output" =~ "t1" ]] || false
+}
+
+@test "fetch: fetch main" {
```
Comment (macneale4) [inline integration-tests/bats/fetch.bats:68]: This test isn't valid is there isn't a check for the absence of something. You should create another branch on the remote, and verify that it doesn't get fetched when you just fetch main.

File: integration-tests/bats/fetch.bats [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+    dolt init
+    dolt remote add origin file://../rem1
+    dolt remote add test-remote file://../rem1
+    dolt push origin main
+
+    cd $TMPDIRS
+    dolt clone file://rem1 repo2
+    cd $TMPDIRS/repo2
+    dolt branch feature
+    dolt remote add test-remote file://../rem1
```
Comment (macneale4) [inline integration-tests/bats/fetch.bats:20]: This remote should point to a remote that has different content. Then in tests which pull from `test-remote`, you can verify that you didn't fetch from origin.

**Review by macneale4 — CHANGES_REQUESTED** [review]
Lookin' good. A few minor things

File: integration-tests/bats/fetch.bats [hunk 6]
```diff
@@ -0,0 +1,342 @@
...
+
+@test "fetch: basic fetch" {
+    cd repo2
+    dolt fetch
+
+    run dolt diff main origin/main
+    [ "$status" -eq 0 ]
+    [[ "$output" =~ "added table" ]] || false
+
+    run dolt sql -q "show tables as of hashof('origin/main')" -r csv
```
Comment (zachmu) [inline integration-tests/bats/fetch.bats:48]: can just say `as of 'origin/main'` here

File: integration-tests/bats/fetch.bats [hunk 7]
```diff
@@ -0,0 +1,342 @@
...
+    [[ "$output" =~ "t1" ]] || false
+}
+
+@test "fetch: fetch specific ref" {
+    cd repo2
+    dolt fetch test-remote refs/heads/main:refs/remotes/test-remote/main
+
+    run dolt diff main test-remote/main
+    [ "$status" -eq 0 ]
+    [[ "$output" =~ "added table" ]] || false
```
Comment (zachmu) [inline integration-tests/bats/fetch.bats:114]: This test has the same problem Neil pointed out: the assertions here look exactly the same as origin.

You need different contents in the thing you're fetching so you can be sure you got the right thing

File: integration-tests/bats/fetch.bats [hunk 8]
```diff
@@ -0,0 +1,342 @@
...
+    cd ../repo2
+    dolt fetch origin feature
+
+    run dolt diff main origin/feature
+    [ "$status" -eq 0 ]
+    [[ "$output" =~ "added table" ]] || false
+
+    run dolt sql -q "show tables as of hashof('origin/feature')" -r csv
+    [ "${#lines[@]}" -eq 2 ]
+    [[ "$output" =~ "Table" ]] || false
```
Comment (zachmu) [inline integration-tests/bats/fetch.bats:135]: Same comment

It's not enough to just push / fetch different refs, the assertions need to rule out that you fetched a different one than you intended

Comment (zachmu) [reply integration-tests/bats/fetch.bats:135]: Same applies to all the tests below. Put a new commit / table on the thing you're fetching so you know you got the right thing

**Review by zachmu — APPROVED** [review]
This looks pretty good, but the tests still need some work. You should have a different set of tables on the different remotes, or some other way to assert that you fetched the right one. Once you fix that you're good to go.

**Pull request merged**